flowmachine/flowmachine/core/cache.py

```diff
@@ -0,0 +1,246 @@
+              LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
+              WHERE relkind = 'r' AND relname='{table_name}' AND nspname='{table_schema}'
+    """
+    try:
+        return int(connection.fetch(sql)[0][0])
+    except IndexError:
+        raise ValueError(
+            f"Table '{table_schema}.{table_name}' does not exist on this connection."
+        )
+
+
+def size_of_cache(connection: "Connection") -> int:
+    """
+    Get the total size in bytes of all cache tables.
+
+    Parameters
+    ----------
+    connection : "Connection"
+
+    Returns
+    -------
+    int
+        Number of bytes in total used by cache tables
+
+    """
+    sql = """SELECT sum(pg_total_relation_size(c.oid)) as total_bytes 
+        FROM pg_class c 
+            LEFT JOIN pg_namespace n 
+        ON n.oid=c.relnamespace 
+        INNER JOIN cache.cached ON
+         relname=cached.tablename AND nspname=cached.schema 
+        WHERE NOT cached.class='Table'"""
+    cache_bytes = connection.fetch(sql)[0][0]
+    return 0 if cache_bytes is None else int(cache_bytes)
+
+
+def compute_time(connection: "Connection", query_id: str) -> float:
+    """
+    Get the time in ms that a cached query took to compute.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    query_id : str
+        md5 identifier of the query
+
+    Returns
+    -------
+    float
+        Number of seconds the query took to compute
+
+    """
+    try:
+        return float(
+            connection.fetch(
+                f"SELECT compute_time FROM cache.cached WHERE query_id='{query_id}'"
+            )[0][0]
+            / 1000
+        )
+    except IndexError:
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def score(connection: "Connection", query_id: str) -> float:
+    """
+    Get the current cache score for a cached query.
+
+    Parameters
+    ----------
+    connection: "Connection"
+    query_id : str
+        md5 id of the cached query
+
+    Returns
+    -------
+    float
+        Current cache score of this query
+
+    """
+    try:
+        return float(
+            connection.fetch(
+                f"SELECT cache_score FROM cache.cached WHERE query_id='{query_id}'"
+            )[0][0]
+        )
+    except IndexError:
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def rescore(connection: "Connection", query: "Query", half_life: float) -> float:
+    """
+    Calculate a new cache score for a cached query object.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    query : "Query"
+        "Query" object to score
+    half_life : float
+        Memory decay halflife. Smaller values will decay more slowly.
+
+    Returns
+    -------
+    float
+        Updated cache score
+    """
+    byte_size = size_of_table(connection, *query.table_name.split(".")[::-1])
+    runtime = compute_time(connection, query.md5)
```

Comment:
runtime = max(some_value_large_enough_for_stuff_to_stick_around, compute_time(connection, query.md5))?

Reply:
I don't think that quite resolves it though.. Because the score grows, you'd need `some_value_large_enough_for_stuff_to_stick_around` to change to reflect that, and you end up giving stuff which arguably _shouldn't_ be cached an artificially high score, thereby keeping it around when it should vanish.